pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ConnectionsLimiter.java

```diff
@@ -0,0 +1,102 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.broker.service;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.mutable.MutableInt;
+import org.apache.pulsar.broker.ServiceConfiguration;
+
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.regex.Pattern;
+
+@Slf4j
+public class ConnectionsLimiter {
+
+    private final int maxConnectionPerIp;
+    private final int maxConnections;
+
+    private final static Map<String, MutableInt> CONNECTIONS = new HashMap<>();
+    private final static MutableInt CURRENT_CONNECTION_NUM = new MutableInt(0);
+    private final static Pattern IPV4_PATTERN = Pattern
+            .compile("^" + "(25[0-5]|2[0-4]\\d|[0-1]?\\d?\\d)" + "(\\.(25[0-5]|2[0-4]\\d|[0-1]?\\d?\\d)){3}" + "$");
+
+    private final static ReentrantLock lock = new ReentrantLock();
+
+    public ConnectionsLimiter(ServiceConfiguration configuration) {
+        maxConnections = configuration.getBrokerMaxConnections();
+        maxConnectionPerIp = configuration.getBrokerMaxConnectionsPerIp();
+    }
+
+    public boolean increaseConnection(SocketAddress remoteAddress) {
+        if (maxConnectionPerIp <= 0 && maxConnections <= 0) {
+            return true;
+        }
+        if (!(remoteAddress instanceof InetSocketAddress) || !IPV4_PATTERN.matcher(
+                ((InetSocketAddress) remoteAddress).getHostString()).find()) {
+            return true;
+        }
+        lock.lock();
+        try {
+            String ip = ((InetSocketAddress) remoteAddress).getHostString();
+            CONNECTIONS.putIfAbsent(ip, new MutableInt(0));
+            CONNECTIONS.get(ip).increment();
+            CURRENT_CONNECTION_NUM.increment();
+            if (CURRENT_CONNECTION_NUM.getValue() > maxConnections
+                    || CONNECTIONS.get(ip).getValue() > maxConnectionPerIp) {
+                return false;
```

Comment:
If the connection is rejected, the number of `CURRENT_CONNECTION_NUM` and `CONNECTIONS.get(ip)` should be decrement?

Reply:
Of course, they are all increased before returning false